src/LanguageServer/Impl/Indexing/IndexManager.cs

```diff
@@ -0,0 +1,156 @@
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Python.Analysis.Core.Interpreter;
+using Microsoft.Python.Analysis.Documents;
+using Microsoft.Python.Core;
+using Microsoft.Python.Core.Diagnostics;
+using Microsoft.Python.Core.Idle;
+using Microsoft.Python.Core.IO;
+using Microsoft.Python.Parsing;
+
+namespace Microsoft.Python.LanguageServer.Indexing {
+    internal class IndexManager : IIndexManager {
+        private static int DefaultReIndexDelay = 350;
+        private readonly ISymbolIndex _symbolIndex;
+        private readonly IFileSystem _fileSystem;
+        private readonly string _workspaceRootPath;
+        private readonly string[] _includeFiles;
+        private readonly string[] _excludeFiles;
+        private readonly IIdleTimeService _idleTimeService;
+        private readonly CancellationTokenSource _allIndexCts = new CancellationTokenSource();
+        private readonly TaskCompletionSource<bool> _addRootTcs = new TaskCompletionSource<bool>();
+        private readonly HashSet<IDocument> _pendingDocs = new HashSet<IDocument>(new UriDocumentComparer());
+        private readonly PythonLanguageVersion _version;
+        private DateTime _lastPendingDocAddedTime;
+
+        public IndexManager(ISymbolIndex symbolIndex, IFileSystem fileSystem, PythonLanguageVersion version, string rootPath, string[] includeFiles,
+            string[] excludeFiles, IIdleTimeService idleTimeService) {
+            Check.ArgumentNotNull(nameof(fileSystem), fileSystem);
+            Check.ArgumentNotNull(nameof(rootPath), rootPath);
+            Check.ArgumentNotNull(nameof(includeFiles), includeFiles);
+            Check.ArgumentNotNull(nameof(excludeFiles), excludeFiles);
+
+            _symbolIndex = symbolIndex;
+            _fileSystem = fileSystem;
+            _workspaceRootPath = rootPath;
+            _includeFiles = includeFiles;
+            _excludeFiles = excludeFiles;
+            _idleTimeService = idleTimeService;
+            _idleTimeService.Idle += OnIdle;
+            _version = version;
+            ReIndexingDelay = DefaultReIndexDelay;
+            StartAddRootDir();
+        }
+
+        public int ReIndexingDelay { get; set; }
+
+        private void StartAddRootDir() {
+            foreach (var fileInfo in WorkspaceFiles()) {
+                if (ModulePath.IsPythonSourceFile(fileInfo.FullName)) {
```

Comment:
https://github.com/Microsoft/python-language-server/issues/637, for the reference